## Previouis Findings 
* There are 10k rows and 8 columns of data
* About 8% of the data is missing, which is not a lot and can be fixed by dropping or imputing the missing values.
* "Payment Method" and "Location" columns have the most missing data, and this is needed to be handled carefully.
* All data types of are in str data type, which needs to be converted to appropriate data types for analysis. 
* Column names should be in snake_format for easy access and readability.
* Missing values are not only in NaN format but some are in texts like "ERROR", "UNKNOWN" etc., which need to be handled separately and very carefully.
* If we count the number of wrong vales, like "UNKNOWN" or "ERROR" in Payment and Location columns, and add it with number of null rows, the number of actual missing data adds up more than just the null data, thereby there needs to be handled carefully or can be dropped in case of confusion since the percentage is comparatively low enough from the total data. 
* A somewhat promision soltution that can be applies to the above problem would be imputing the wrong values and missing values with mode.
* For the missing values of the Item column, we chould drop the rows with respect to null values of that column since we dont know the menu to find out the missing items, and the null data percentage is very low enough to take this decision I guess.

# Imports

In [1]:
import pandas as pd
import numpy as np

# Data Import

In [2]:
df = pd.read_csv("dirty_cafe_sales.csv")
df.head()

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11


In [3]:
df.columns = df.columns.str.lower().str.replace(" ", "_")
df = df.drop(columns="transaction_id")  # Drop transaction_id column as it is not needed asuming this data will be fed into some ML model
df.head()

,item,quantity,price_per_unit,total_spent,payment_method,location,transaction_date
0,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11


In [4]:
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   item              9667 non-null   str  
 1   quantity          9862 non-null   str  
 2   price_per_unit    9821 non-null   str  
 3   total_spent       9827 non-null   str  
 4   payment_method    7421 non-null   str  
 5   location          6735 non-null   str  
 6   transaction_date  9841 non-null   str  
dtypes: str(7)
memory usage: 3.5 MB


In [5]:
print(f"Missing Value Counts\n{df.isnull().sum()}\n")
print(f"Missing Value Counts : Percentage\n{df.isnull().sum() / len(df) * 100}\n")

Missing Value Counts
item                 333
quantity             138
price_per_unit       179
total_spent          173
payment_method      2579
location            3265
transaction_date     159
dtype: int64

Missing Value Counts : Percentage
item                 3.33
quantity             1.38
price_per_unit       1.79
total_spent          1.73
payment_method      25.79
location            32.65
transaction_date     1.59
dtype: float64



In [6]:
print(f"Total Missing Value Counts\n{df.isnull().sum().sum()} Out of {len(df)*len(df.columns)}\n")
print(f"Total Missing Value Counts : Percentage : {df.isnull().sum().sum() / (len(df)*len(df.columns)) * 100}\n")

Total Missing Value Counts
6826 Out of 70000

Total Missing Value Counts : Percentage : 9.751428571428573



# Null conversion

In [7]:
df.sample(10)

,item,quantity,price_per_unit,total_spent,payment_method,location,transaction_date
8726,Cookie,1,1.0,1.0,NaN,In-store,2023-02-11
1429,Coffee,NaN,2.0,6.0,NaN,Takeaway,2023-12-19
2621,Tea,1,1.5,1.5,Cash,NaN,2023-08-26
4186,Cake,3,3.0,9.0,Cash,UNKNOWN,2023-03-13
2598,Tea,3,1.5,4.5,Digital Wallet,Takeaway,2023-07-24
1300,Cake,4,3.0,12.0,Credit Card,NaN,2023-11-04
7105,Tea,5,1.5,7.5,NaN,In-store,2023-11-26
2567,Sandwich,5,4.0,20.0,ERROR,NaN,2023-12-22
7561,Tea,1,1.5,1.5,Cash,In-store,2023-09-21
6983,Salad,4,5.0,20.0,Credit Card,Takeaway,2023-08-26


In [8]:
df.columns

Index(['item', 'quantity', 'price_per_unit', 'total_spent', 'payment_method',
       'location', 'transaction_date'],
      dtype='str')

In [9]:
error_set = ['UNKNOWN', 'ERROR', None]

In [10]:
df = df.replace(error_set, np.nan)

In [11]:
df[["quantity","price_per_unit","total_spent"]] = df[["quantity","price_per_unit","total_spent"]].astype("float32")

In [12]:
df.sample(10)

,item,quantity,price_per_unit,total_spent,payment_method,location,transaction_date
8130,Sandwich,NaN,4.0,12.0,Digital Wallet,Takeaway,2023-01-29
2179,Smoothie,1.0,4.0,4.0,Credit Card,Takeaway,2023-06-28
9594,Juice,5.0,3.0,15.0,Cash,In-store,2023-07-23
7952,Tea,5.0,1.5,7.5,NaN,NaN,2023-03-15
9948,Tea,1.0,1.5,1.5,NaN,In-store,2023-09-27
1904,Coffee,1.0,2.0,2.0,NaN,Takeaway,2023-05-18
5471,Cake,5.0,3.0,15.0,NaN,Takeaway,NaN
7438,Juice,1.0,3.0,3.0,Digital Wallet,NaN,2023-05-08
3306,Coffee,4.0,2.0,8.0,Digital Wallet,NaN,2023-04-03
7788,Salad,5.0,5.0,25.0,Digital Wallet,In-store,2023-02-27


In [13]:
print(f"Missing Value Counts\n{df.isnull().sum()}\n")

Missing Value Counts
item                 969
quantity             479
price_per_unit       533
total_spent          502
payment_method      3178
location            3961
transaction_date     460
dtype: int64



All unwanted values are converted to NaN for better access and simplification of data processing process.

# Recovering `quantity`, `price_per_unit` and `total_spent` column data 

In [14]:
df['quantity'] = df['quantity'].fillna(df['total_spent'] // df['price_per_unit'])

df['price_per_unit'] = df['price_per_unit'].fillna(df['total_spent'] / df['quantity'])

df['total_spent'] = df['total_spent'].fillna(df['quantity'] * df['price_per_unit'])

In [15]:
df.sample(10)

,item,quantity,price_per_unit,total_spent,payment_method,location,transaction_date
2578,NaN,5.0,5.0,25.0,Cash,NaN,2023-08-04
5348,Cookie,4.0,1.0,4.0,NaN,In-store,2023-01-11
4208,Salad,1.0,5.0,5.0,Cash,Takeaway,2023-01-05
3911,Cookie,5.0,1.0,5.0,Credit Card,NaN,2023-08-05
6832,Coffee,1.0,2.0,2.0,Credit Card,Takeaway,2023-09-29
2715,Coffee,2.0,2.0,4.0,NaN,NaN,2023-05-18
9352,Salad,5.0,5.0,25.0,NaN,In-store,2023-08-31
5313,Coffee,5.0,2.0,10.0,NaN,In-store,2023-01-05
4178,Cookie,5.0,1.0,5.0,NaN,Takeaway,2023-06-04
4254,Smoothie,3.0,4.0,12.0,NaN,NaN,2023-12-17


In [16]:
print(f"Missing Value Counts\n{df.isnull().sum()}\n")

Missing Value Counts
item                 969
quantity              38
price_per_unit        38
total_spent           40
payment_method      3178
location            3961
transaction_date     460
dtype: int64



# Recovering `item` column data from previously recovered columns

In [17]:
price_map = df.drop_duplicates('item').set_index('item')['price_per_unit'].to_dict()
del price_map[np.nan]
price_map

{'Coffee': 2.0,
 'Cake': 3.0,
 'Cookie': 1.0,
 'Salad': 5.0,
 'Smoothie': 4.0,
 'Sandwich': 4.0,
 'Juice': 3.0,
 'Tea': 1.5}

In [18]:
item_lookup = {v: k for k, v in price_map.items() if pd.notna(k)}
item_lookup

{2.0: 'Coffee',
 3.0: 'Juice',
 1.0: 'Cookie',
 5.0: 'Salad',
 4.0: 'Sandwich',
 1.5: 'Tea'}

In [19]:
df['item'] = df['item'].fillna(df['price_per_unit'].map(item_lookup))

In [20]:
print(f"Missing Value Counts\n{df.isnull().sum()}\n")

Missing Value Counts
item                   6
quantity              38
price_per_unit        38
total_spent           40
payment_method      3178
location            3961
transaction_date     460
dtype: int64



We can see a drastic change in the number of recovered data, from 969 missing data in `item` column to now 6. <br>
The 2-4no. columns could be again checked for data recovery but that part is being skipped and the rows will be dropped for simplicity, since that is very little data to waste time on.

# Imputing the last three columns

In [21]:
df['payment_method'] = df['payment_method'].fillna(df['payment_method'].mode()[0])
df['location'] = df['location'].fillna(df['location'].mode()[0])
df['transaction_date'] = df['transaction_date'].ffill() 
df['transaction_date'] = pd.to_datetime(df['transaction_date'])

In [22]:
df.dropna(axis="rows", inplace=True)
# Converting the last columns datatype and sorting based on that column
df = df.sort_values(by='transaction_date', ascending=True)
df = df.reset_index(drop=True)
df['quantity'] = df['quantity'].astype("int16")
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 9942 entries, 0 to 9941
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   item              9942 non-null   str           
 1   quantity          9942 non-null   int16         
 2   price_per_unit    9942 non-null   float32       
 3   total_spent       9942 non-null   float32       
 4   payment_method    9942 non-null   str           
 5   location          9942 non-null   str           
 6   transaction_date  9942 non-null   datetime64[us]
dtypes: datetime64[us](1), float32(2), int16(1), str(3)
memory usage: 1.8 MB


In [23]:
df.sample(10)

,item,quantity,price_per_unit,total_spent,payment_method,location,transaction_date
6604,Cookie,5,1.0,5.0,Digital Wallet,Takeaway,2023-09-01
8640,Coffee,1,2.0,2.0,Digital Wallet,Takeaway,2023-11-13
7552,Cake,5,3.0,15.0,Digital Wallet,Takeaway,2023-10-06
9515,Cake,3,3.0,9.0,Digital Wallet,Takeaway,2023-12-16
4254,Cookie,1,1.0,1.0,Digital Wallet,In-store,2023-06-07
1794,Smoothie,5,4.0,20.0,Credit Card,Takeaway,2023-03-07
5307,Sandwich,1,4.0,4.0,Cash,Takeaway,2023-07-15
2400,Sandwich,4,4.0,16.0,Cash,In-store,2023-03-29
768,Coffee,1,2.0,2.0,Digital Wallet,Takeaway,2023-01-29
2368,Tea,2,1.5,3.0,Digital Wallet,In-store,2023-03-28


# Exporting Data

In [24]:
df.to_csv("cleaned_cafe_sales_data.csv",index=False)

# Conclusion
The data cleaning process has been successfully completed.
* The unnecessary columns are removed.
* Most of the missing data is recovered back without using any imputation or such, but simple mathematics of `item` : `quantity` *`price_per_unit` =`total_spent`.
* Imputation like mode and forward fill were done only on the last three categorical columns.
* The `transaction_date` column is kept for any further analysis else if this data is going to be fedd into some ML model, then it should be dropped.
* The new cleaned data with about only 0.5% loss which before cleaning was about 9.7% is imported successfully.